### PR TITLE
ci: refactor curl / wget test helpers with retries

### LINKED ISSUE #11994: CI: Refactor curl / wget test helpers with retries
It'd be nice if we have a test helper wrapper for `curl` and `wget` which has retries built in. These retries should only be applied to commands that are expected to be used for connection with the outside / world.

For example:

```
testhelpers.CurlWithRetries(..., retries int, ...)
```

This would eliminate the need for one-off PRs like https://github.com/cilium/cilium/pull/11797 and https://github.com/cilium/cilium/pull/11993. These types of PRs are prone to accumulate tech debt as the number of retries could be incosistent across the code base, etc.. Also, issues like missing a place where retries would be needed, could be eliminated.

### PR DESCRIPTION
**Pull request by LexCC**
Fixes: #11994

```release-note
ci: refactor curl / wget test helpers with retries
```

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

File: test/helpers/wrappers.go
Comment (nebril) [inline test/helpers/wrappers.go]: I think it makes a lot of sense to refactor this, but I feel that this functions shows too many implementation details and nesting `CurlFail` into `CurlWithRetries` is not necessary.

I would like `CurlWithRetries` to accept the same arguments as `CurlFail` with addition of `--retry` flag.

I think this function should call `CurlFail` and then perform string replacement on it that would replace `curl` with `curl --retry %d`.

**Review by nebril — CHANGES_REQUESTED** [review]
Thanks a lot for the PR, it was bugging me for quite some time now.

I left a comment inline, please consider changing it as requested.

**Review by christarazi — COMMENTED** [review]
Thanks for the PR! Overall I think this is good. +1 to @nebril's idea to refactor some of the details away in the new function

File: test/helpers/wrappers.go
Comment (LexCC) [reply test/helpers/wrappers.go]: Thanks for reviewing. I think the function will become something like this:
``CurlWithRetries`` can use ``failed`` flag to distinguish CurlFail and non-CurlFail.
```
func CurlWithRetries(endpoint string, retries int, failed bool, optionalValues ...interface{}) string {
	if failed {
		return fmt.Sprintf(
			`%s --retry %d`,
			CurlFail(endpoint, optionalValues...), retries)
	}
	if len(optionalValues) > 0 {
		endpoint = fmt.Sprintf(endpoint, optionalValues...)
	}
	return fmt.Sprintf(
		`curl --path-as-is -s  -D /dev/stderr --output /dev/stderr --retry %d %s`,
		retries, endpoint)
}
```
That made this line https://github.com/cilium/cilium/pull/12408/files#diff-0ac1702ef92e08f680c4d7d79842e042R95 can reuse the funtion.

Comment (nebril) [reply test/helpers/wrappers.go]: Looks good. One nit on argument names - `failed` arg should be called `fail`, as this is equivalent of curl `--fail` flag.

Also please make sure that adding `--retry %d` at the end of command always works as intended (although we will probably figure this out when running CI).

Comment (LexCC) [reply test/helpers/wrappers.go]: Updated! Please take a look.

File: test/helpers/wrappers.go
Comment (christarazi) [inline test/helpers/wrappers.go]: Nit:
```suggestion
// If flag "fail" is true, the function will call CurlFail() and add --retry flag at the end of the command and return.
```

File: test/helpers/wrappers.go
Comment (christarazi) [inline test/helpers/wrappers.go]: Nit:

```suggestion
// CurlWithRetries returns the string representation of the curl command that
// retries the request if transient problems occur. The parameter "retries"  indicates the maximum number of attempts.
```

Also, could you reformat this comment to be consistent with the rest of the file? This comment block is wider than the rest (`gq` if you're using vim ;) )

**Review by christarazi — APPROVED** [review]
LGTM; just a few minor comments, thanks. I think it's worth mentioning that there a few calls that can be simplified (they pass a format string as the 1st arg and utilize `optionalValues` to fill in), however we can leave that to a follow up.

File: test/helpers/wrappers.go
Comment (LexCC) [reply test/helpers/wrappers.go]: Thanks! I've updated the PR. Please take a look.

**Comment by christarazi** [comment]
test-me-please

**Comment by JieJhih** [comment]
test-me-please

**Comment by christarazi** [comment]
@JieJhih Only members with commit access can trigger CI. If you need to trigger CI in the future, feel free to reach out to me here on or Slack for quicker turnaround time. I'll trigger it now.

**Comment by christarazi** [comment]
test-me-please

**Comment by JieJhih** [comment]
> @JieJhih Only members with commit access can trigger CI. If you need to trigger CI in the future, feel free to reach out to me here or Slack for quicker turnaround time. I'll trigger it now.

Thanks! I just wonder why Cilium-Ginkgo-GKE unit test didn't passed, I'm not sure if that is related to this PR.

**Comment by christarazi** [comment]
> > @JieJhih Only members with commit access can trigger CI. If you need to trigger CI in the future, feel free to reach out to me here or Slack for quicker turnaround time. I'll trigger it now.
> 
> Thanks! I just wonder why Cilium-Ginkgo-GKE unit test didn't passed, I'm not sure if that is related to this PR.

We have known issues with tests running on GKE. I would consider this PR as passing.

**Review by nebril — APPROVED** [review]

**Review by qmonnet — APPROVED** [review]
Thanks!

**Comment by christarazi** [comment]
Looks like this PR had two identical commits that should have been squashed. We should be on the look out for issues like this next time before merging.

```
$ git log
...
* 8f0b57d0b kvstore: Add metric to count quorum errors
* a33605936 ci: refactor curl / wget test helpers with retries
* 21be15ebe ci: refactor curl / wget test helpers with retries
* 38d9bf589 bpf: explicitly set ttl in tunnel key
```

**Comment by pchaigno** [comment]
Marking for v1.8 backports to ease the backport of #13694.

**Pull request merged**
